registration/registration.go

```diff
@@ -0,0 +1,362 @@
+		return err
+	}
+	defer round.Close()
+
+	logging.FromContext(ctx).Info("found round in progress, scheduling it", zap.Uint("epoch", round.epoch))
+	root, err := round.calcMembershipRoot()
+	if err != nil {
+		return fmt.Errorf("calculating membership root: %w", err)
+	}
+	if err := r.workerSvc.ExecuteRound(ctx, round.epoch, root); err != nil {
+		return fmt.Errorf("scheduling in-progress round for epoch %d: %w", round.epoch, err)
+	}
+	return nil
+}
+
+func (r *Registration) onNewProof(ctx context.Context, proof shared.NIP) error {
+	logger := logging.FromContext(ctx).Named("on-proof").With(zap.Uint("epoch", proof.Epoch))
+	logger.Info("received new proof", zap.Uint64("leaves", proof.Leaves))
+
+	// Retrieve the list of round members for the round.
+	// This is temporary until we remove the list of members from the proof.
+	opts := append(r.newRoundOpts(), withFailfIfNotExists())
+	round, err := newRound(proof.Epoch, r.dbdir, opts...)
+	switch {
+	case errors.Is(err, fs.ErrNotExist):
+		return nil
+	case err != nil:
+		return err
+	}
+
+	members := round.getMembers()
+	if err := round.Close(); err != nil {
+		logger.Error("failed to close round", zap.Error(err))
+	}
+	if err := r.db.SaveProof(ctx, proof, members); err != nil {
+		return fmt.Errorf("saving proof in DB: %w", err)
+	} else {
+		logger.Info("proof saved in DB", zap.Int("members", len(members)), zap.Uint64("leaves", proof.Leaves))
+	}
+
+	// Rotate Proof of Work challenge.
+	params := r.powVerifiers.Params()
+	params.Challenge = proof.Root
+	r.powVerifiers.SetParams(params)
+	membersMetric.DeleteLabelValues(epochToRoundId(proof.Epoch))
+	return nil
+}
+
+func (r *Registration) newRoundOpts() []newRoundOptionFunc {
+	return []newRoundOptionFunc{
+		withMaxMembers(r.cfg.MaxRoundMembers),
+		withMaxSubmitBatchSize(r.cfg.MaxSubmitBatchSize),
+		withSubmitFlushInterval(r.cfg.SubmitFlushInterval),
+	}
+}
+
+func (r *Registration) Submit(
+	ctx context.Context,
+	challenge, nodeID []byte,
+	nonce uint64,
+	powParams PowParams,
+) (epoch uint, roundEnd time.Time, err error) {
+	logger := logging.FromContext(ctx)
+
+	err = r.powVerifiers.VerifyWithParams(challenge, nodeID, nonce, powParams)
+	if err != nil {
+		logger.Debug("challenge verification failed", zap.Error(err))
+		return 0, time.Time{}, err
+	}
+	logger.Debug("verified challenge", zap.String("node_id", hex.EncodeToString(nodeID)))
+
+	r.openRoundMutex.RLock()
+	defer r.openRoundMutex.RUnlock()
+	epoch = r.openRound.epoch
+	done, err := r.openRound.submit(ctx, nodeID, challenge)
+	switch {
+	case err == nil:
+		logger.Debug("async-submitted challenge for round", zap.Uint("round", epoch))
+		// wait for actually submitted
+		select {
+		case <-ctx.Done():
+			logger.Debug("context canceled while waiting for challenge to be submitted")
+			return 0, time.Time{}, ctx.Err()
+		case err := <-done:
+			if err != nil {
+				logger.Warn("challenge registration failed", zap.Error(err))
+				return 0, time.Time{}, err
+			}
+		}
```

Comment:
This seems incorrect. Why is `r.openRound.submit` not just a blocking call that returns an error if it fails?

```suggestion
	err := r.openRound.submit(ctx, nodeID, challenge)
	if err != nil {
		logger.Warn("challenge registration failed", zap.Error(err))
		return 0, time.Time{}, err
	}
```

The `select` on the context while waiting for a possible `error` on `done` in my opinion belongs into the `submit` of the `openRound`?

Calls to `submit` are mutually exclusive anyway, so I see no reason to just make this a blocking call?

Reply:
The `round::submit()` is async because I don't want to hold the **shared** (`RLock()`) lock of `openRoundMutex` until the batch is flushed. It's enough for the submit to be batched and the lock can be released (btw, the `openRoundMutex.Unlock()` is missing which is a bug - fixing right now).

When the round is about to be closed, the `openRoundMutex` is obtained **exclusively** (`Lock()`), and the pending batch is flushed (completing all pending `Registration::Submit()`). The round is executed and a new round is opened.

💡 A blocked call to `RWMutex::Lock()` prevents subsequent `RWMutex::RLock()` from acquiring the lock even if it is already held for read by other goroutines, which prevents starvation of the main `Registration` thread that tries to close the open round (https://pkg.go.dev/sync#RWMutex).

Reply:
What I mean is the `Submit` method in `registration.go` could look like this:

```go
func (r *Registration) Submit(
	ctx context.Context,
	challenge, nodeID []byte,
	nonce uint64,
	powParams PowParams,
) (epoch uint, roundEnd time.Time, err error) {
	logger := logging.FromContext(ctx)

	err = r.powVerifiers.VerifyWithParams(challenge, nodeID, nonce, powParams)
	if err != nil {
		logger.Debug("challenge verification failed", zap.Error(err))
		return 0, time.Time{}, err
	}
	logger.Debug("verified challenge", zap.String("node_id", hex.EncodeToString(nodeID)))

	r.openRoundMutex.RLock()
	epoch = r.openRound.epoch
	r.openRoundMutex.RUnlock()
	err = r.openRound.submit(ctx, nodeID, challenge)

	switch {
	case err == nil:
	case errors.Is(err, ErrChallengeAlreadySubmitted):
	case err != nil:
		return 0, time.Time{}, err
	}

	return epoch, r.roundCfg.RoundEnd(r.genesis, epoch), nil
}
```

If we update the `submit` method to:

```go
func (r *round) submit(ctx context.Context, key, challenge []byte) error {
	done, err := r.submitLocked(ctx, key, challenge)
	if err != nil {
		return err
	}

	select {
	case <-ctx.Done():
		return ctx.Err()
	case err := <-done:
		return err
	}
}

func (r *round) submitLocked(ctx context.Context, key, challenge []byte) (<-chan error, error) {
	r.batchMutex.Lock()
	defer r.batchMutex.Unlock()

	if r.members+len(r.pendingSubmits) >= r.maxMembers {
		return nil, ErrMaxMembersReached
	}

	pending, ok := r.pendingSubmits[string(key)]
	if ok {
		switch {
		case bytes.Equal(challenge, pending.challenge):
			return nil, fmt.Errorf("%w: key: %X", ErrChallengeAlreadySubmitted, key)
		default:
			return nil, ErrConflictingRegistration
		}
	}

	if r.batch == nil {
		r.batch = leveldb.MakeBatch(r.maxBatchSize)
		time.AfterFunc(r.flushInterval, r.flushPendingSubmits)
	}

	registered, err := r.db.Get(key, nil)
	switch {
	case errors.Is(err, leveldb.ErrNotFound):

	// OK - challenge is not registered yet.
	case err != nil:
		return nil, fmt.Errorf("failed to check if challenge is registered: %w", err)
	case bytes.Equal(challenge, registered):
		return nil, fmt.Errorf("%w: key: %X", ErrChallengeAlreadySubmitted, key)
	default:
		return nil, ErrConflictingRegistration
	}

	r.batch.Put(key, challenge)
	done := make(chan error, 1)
	r.pendingSubmits[string(key)] = pendingSubmit{
		done,
		challenge,
	}

	if r.batch.Len() >= r.maxBatchSize {
		r.flushPendingSubmitsLocked()
	}
	return done, nil
}
```

`submitLocked` is basically the same as the current `submit` method but alleviates the caller from having to check 2 different errors (`error` & `done`). Simplifying the call in `registration.go`.

Reply:
The `Registration::Submit()` is called concurrently with `Registration::closeRound()`. The proposed code is racy and could lead to submitting to a closed round or writing batch to a closed DB etc.

Reply:
I disagree, if the code wasn't racy before it also isn't racy now? Nothing changed. Calling `submit` does the same thing as before, just that the `done` channel is read from before returning instead of giving the caller that responsibility?

Reply:
It is racy. The code wasn't racy before (that I know of). The challenge used to be batched with `r.openRoundMutex` Rlocked which prevented closing the round. Now this is not true. Consider:
1. `Registration::Submit(challengeA)` is called, executes until right after Unlock:
    ```go
    r.openRoundMutex.RUnlock()
    // halts here
    err = openRound.submit(ctx, nodeID, challenge) // note: removed `r.`, I assume you meant it this way.
    ```
2. `r.closeRound()` is called. It can take `r.openRoundMutex.Lock()`
3. The root is calculated `root, err := r.openRound.calcMembershipRoot()`. It doesn't contain `challengeA`
5. The `challengeA` submit (1.) continues, writes the challenge into a batch too late as the root is already calculated
6. the submit finishes and falsely returns OK to the user